### PR TITLE
Rework the core of class parsing/runtime semantic

### PR DESCRIPTION
**Pull request by rerobika**
Changes:
 - Use the pre-scanner to provide information for the parser about the existence of the class constructor
 - The allocation of the super declarative environment is no longer needed
 - The VM frame context holds the information about the this binding status
 - Reduce the number of class related VM/CBC instructions

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 1]
```diff
@@ -1162,6 +931,101 @@ ecma_op_function_call_external (ecma_object_t *func_obj_p, /**< Function object
...
+/**
+ * Append the bound arguments into the given collection
+ *
+ * Note:
+ *       - The whole bound chain is resolved
+ *       - The first element of the collection contains the bounded this value
+ *
+ * @return target function of the bound function
+ */
+static ecma_object_t * JERRY_ATTR_NOINLINE
```
Comment (ossy-szeged) [inline jerry-core/ecma/operations/ecma-function-object.c]: JERRY_ATTR_NOINLINE should be placed at the beginning, can't be after * character.
See #3347 for details, we ran into this strange windows build fail previously.

File: jerry-core/parser/js/js-parser-expr.c [hunk 2]
```diff
@@ -1809,6 +1775,13 @@ parser_process_unary_expression (parser_context_t *context_p, /**< context */
           context_p->last_cbc_opcode = PARSER_CBC_UNAVAILABLE;
           parser_emit_cbc_literal_from_token (context_p, CBC_PUSH_PROP_THIS_LITERAL);
         }
+#if ENABLED (JERRY_ES2015)
+        else if (context_p->last_cbc_opcode == PARSER_TO_EXT_OPCODE (CBC_EXT_PUSH_SUPER))
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-expr.c]: I think the flush opcode should have a check which disallows flushing this opcode.

File: jerry-core/parser/js/js-parser-internal.h [hunk 3]
```diff
@@ -65,17 +65,14 @@ typedef enum
...
-  PARSER_CLASS_CONSTRUCTOR = (1u << 19),      /**< a class constructor is parsed (this value must be kept in
-                                               *   in sync with ECMA_PARSE_CLASS_CONSTRUCTOR) */
-  PARSER_CLASS_HAS_SUPER = (1u << 20),        /**< class has super reference */
-  PARSER_CLASS_IMPLICIT_SUPER = (1u << 21),   /**< class has implicit parent class */
-  PARSER_CLASS_STATIC_FUNCTION = (1u << 22),  /**< this function is a static class method */
-  PARSER_CLASS_SUPER_PROP_REFERENCE = (1u << 23),  /**< super property call or assignment */
+  /* These three status flags must be in this order. See PARSER_SAVED_FLAGS_OFFSET. */
+  PARSER_CLASS_CONSTRUCTOR = (1u << 19),      /**< a class constructor is parsed */
+  PARSER_ALLOW_SUPER = (1u << 20),            /**< allow super property access */
+  PARSER_ALLOW_SUPER_CALL = (1u << 21),       /**< allow super constructor call */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-internal.h]: comment: PARSER_CLASS_CONSTRUCTOR must be present

File: jerry-core/parser/js/js-parser-internal.h [hunk 4]
```diff
@@ -65,17 +65,14 @@ typedef enum
...
   PARSER_FUNCTION_HAS_REST_PARAM = (1u << 18), /**< function has rest parameter */
-  /* These four status flags must be in this order. See PARSER_CLASS_PARSE_OPTS_OFFSET. */
-  PARSER_CLASS_CONSTRUCTOR = (1u << 19),      /**< a class constructor is parsed (this value must be kept in
-                                               *   in sync with ECMA_PARSE_CLASS_CONSTRUCTOR) */
-  PARSER_CLASS_HAS_SUPER = (1u << 20),        /**< class has super reference */
-  PARSER_CLASS_IMPLICIT_SUPER = (1u << 21),   /**< class has implicit parent class */
-  PARSER_CLASS_STATIC_FUNCTION = (1u << 22),  /**< this function is a static class method */
-  PARSER_CLASS_SUPER_PROP_REFERENCE = (1u << 23),  /**< super property call or assignment */
+  /* These three status flags must be in this order. See PARSER_SAVED_FLAGS_OFFSET. */
+  PARSER_CLASS_CONSTRUCTOR = (1u << 19),      /**< a class constructor is parsed */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-internal.h]: comment: PARSER_ALLOW_SUPER must be also present

File: jerry-core/parser/js/js-scanner.c [hunk 5]
```diff
@@ -2361,6 +2390,23 @@ scanner_scan_all (parser_context_t *context_p, /**< context */
             break;
           }
 
+          if (context_p->token.type == LEXER_LITERAL
+              && LEXER_IS_IDENT_OR_STRING (context_p->token.lit_location.type)
+              && lexer_compare_literal_to_string (context_p, "constructor", 11))
+          {
+            if (stack_top == SCAN_STACK_IMPLICIT_CLASS_CONSTRUCTOR)
+            {
+              lexer_skip_spaces (context_p);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-scanner.c]: I think this is unnecessary.

File: jerry-core/ecma/operations/ecma-function-object.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-function-object.c]: Thank you for the investigation! :)

**Comment by rerobika** [comment]
@zherczeg @ossy-szeged Thanks for the reviews, I've updated the PR according to them.

**Pull request merged**